### DISCUSSION TITLE
Can I make WhenCreated, WhenModified columns nullable?

### DISCUSSION THREAD

**Discussion by Graham-McKenzie** [ebean-orm/ebean#3239] [General]
Hi all.  I want to add the @WhenCreated and similar annotations to a table that already has data, and has no sensible defaults for the timestamped columns.  So ideally I want to use the annotations to populate the auditing columns from now on, but leave existing data null.  Is there a way I can do this?  I've tried e.g.
```
    @WhenModified
    @Column(name = "created_at", nullable = true)
    @DbDefault(value = "now()")
    private OffsetDateTime createdAt;
```
But the generated sql makes the column not null regardless.
Thanks for any advice.

**Comment by rob-bygrave** [comment]
> Is there a way I can do this? 

What we could do is leave the code like:
```java
    @WhenModified
    private OffsetDateTime createdAt;
```
When we generate db migration DDL for this there are 2 things generated being the `*.sql` and the `*.model.xml`.

What we can do is we can manually tweak the generated `*.sql` after it has been generated and not change the `*.model.xml` - and for this case we tweak the generated ddl to allow `created_at` to be nullable (or for example remove entire create table statements for table that already exist and we don't want to create etc). That is, it is allowed and valid to manually edit the generated sql including adding addition sql into the db migration if we need to.

That is, this mechanism used allows a difference between the actual db migration sql and the "logical model" (where the logical model is defined by the `model.xml`).  As we have not changed the `model.xml` then when the next db migration is generated it sees no further difference for the `createdAt` property (the db migration diff works on the difference it detects in the logical model).

In this way we can have differences between the "physical model" (in terms of sql in the db migrations) and the "logical model".  I think it's fair to say this is a feature that can be useful and also can be abused - we don't want to over use this feature.

----
The background to this is that `@WhenCreated` and `@WhenModified` really really want to define NOT NULL and were not designed for this scenario.


Does that make sense?

**Reply by rob-bygrave** [reply]
> @DbDefault(value = "now()")

Note to readers, ebean will never bind a null for `@WhenCreated` and `@WhenModified` ... in the above the `@DbDefault(value = "now()")` is purely there for the generated db migration sql.

**Reply by Graham-McKenzie** [reply]
It does make sense, thank you, and that's what we have done.  I was more wondering whether it's possible to control the code generation in the first place.  But if not, that's fine.  Thanks for the reply.
